README.md
<div align="center">
  <br>
  <img alt="KhulnaSoft OpenSource" src="https://i.ibb.co/7jPXt0Z/logo1-92f1a87f.png" width="300px">
  <h1>🍕 KhulnaSoft OpenSource Insights API 🍕</h1>
  <strong>The path to your next Open Source contribution</strong>
  <br>
  <br>
  <a href="https://www.digitalocean.com/?refcode=c65a90d0956d&utm_campaign=Referral_Invite&utm_medium=Referral_Program&utm_source=badge"><img src="https://web-platforms.sfo2.cdn.digitaloceanspaces.com/WWW/Badge%201.svg" alt="DigitalOcean Referral Badge" /></a>
</div>
<br>
<p align="center">
  <img src="https://img.shields.io/github/languages/code-size/khulnasoft-opensource/api.khulnasoft.com" alt="GitHub code size in bytes">
  <img src="https://img.shields.io/github/commit-activity/w/khulnasoft-opensource/api.khulnasoft.com" alt="GitHub commit activity">
  <a href="https://github.com/khulnasoft-opensource/api.khulnasoft.com/issues">
    <img src="https://img.shields.io/github/issues/khulnasoft-opensource/api.khulnasoft.com" alt="GitHub issues">
  </a>
  <a href="https://github.com/khulnasoft-opensource/api.khulnasoft.com/releases">
    <img src="https://img.shields.io/github/v/release/khulnasoft-opensource/api.khulnasoft.com.svg?style=flat" alt="GitHub Release">
  </a>
  <a href="https://discord.gg/U2peSNf23P">
    <img src="https://img.shields.io/discord/714698561081704529.svg?label=&logo=discord&logoColor=ffffff&color=7389D8&labelColor=6A7EC2" alt="Discord">
  </a>
  <a href="https://twitter.com/Khulnasoft">
    <img src="https://img.shields.io/twitter/follow/Khulnasoft?label=Follow&style=social" alt="Twitter">
  </a>
</p>

## 🚀 Live release environments

- [api.khulnasoft.com](https://api.khulnasoft.com/)
- [beta.api.khulnasoft.com](https://beta.api.khulnasoft.com/)

## 📖 Prerequisites

In order to run the project we need the following software binaries installed on our development machines:

- `node>=16.7.0`
- `npm>=8.0.0`
- `docker>=20.10.12`

## 🖥️ Local development

## 📡 Setting Up Your Supabase Instance

Before setting up your local PostgreSQL database, a Supabase instance should be created. To do this, follow these steps:

**1. Create a new Supabase project:** Visit [Supabase](https://supabase.com/) and create a new project.

**2. Set Environment Variables in .env file:** Once your project is created, Supabase will provide a URL and an API key. Set these in your project's `.env` file:

```
SUPABASE_URL=your_supabase_url
SUPABASE_API_KEY=your_supabase_api_key
SUPABASE_JWT_SECRET=your_supabase_jwt_secret
API_DOMAIN=your_api_domain
```

Replace `your_supabase_url`, `your_supabase_api_key`, `your_supabase_jwt_secret`, and `your_api_domain` with the actual values provided by Supabase and your project's settings.

### 🗄️ Setting Up A PostgreSQL Database Locally

A PostgreSQL Docker container has been set up to facilitate local development. After navigating to the `dev/` directory, this container can be set up as follows:

**1. Obtain SSL Certificates:**  
For secure SSL communication, you need a pair of SSL certificates: `server.crt` and `server.key`. You can generate self-signed certificates by using OpenSSL:

```shell
openssl req -x509 -newkey rsa:4096 -keyout server.key -out server.crt -days 365 -nodes -subj "/CN=localhost"
```

Please note that this generates self-signed certificates which should only be used for local development.

**2.Build the Docker image:**

```shell
docker build -t my_postgres_image -f Dockerfile.local-postgres .
```

This command will build the Docker image using the Dockerfile in the current directory. The previously generated SSL certificates will need to be in the same directory as the Dockerfile.

**3.Run the Docker container:**

```shell
docker run --name my_postgres_container -p 25060:5432 -d my_postgres_image:latest
```

This command will start a new Docker container named my_postgres_container, mapping port 25060 on your local machine to port 5432 on the Docker container.

### 🛠️ Installation

To install the application:

```shell
npm ci
```

To start a local copy of the app on port `3001`:

```shell
npm run start:dev
```

### 🧪 Test

For running the test suite, use the following command. Since the tests run in watch mode by default, some users may encounter errors about too many files being open. In this case, it may be beneficial to [install watchman](https://facebook.github.io/watchman/docs/install.html).

```shell
npm test
```

You can request a coverage report by running the following command:

```shell
npm run test:coverage
```

For writing tests, the rule is move business or service logic to the lib folder and write unit tests. Logic that needs to be in a React component, then leverage tools like [Cypress](https://www.cypress.io/) or [Vitest mocking](https://vitest.dev/guide/mocking.html) to write tests.

### 📦 Docker builds

A development preview can also be run from docker:

```shell
docker build -t api.khulnasoft.com .
docker run -p 8080:3001 api.khulnasoft.com
```

Alternatively you can pull the production container and skip all builds:

```shell
docker run -dit -p 8080:3001 ghcr.io/khulnasoft-opensource/api.khulnasoft.com
```

### 🎨 Code linting

To check the code and styles quality, use the following command:

```shell
npm run lint
```

This will also display during development, but not break on errors.

To fix the linting errors, use the following command:

```shell
npm run format
```

It is advised to run this command before committing or opening a pull request.

### 🕺 OpenAPI Swagger Doc

When the server is run, an OpenAPI swagger doc is generated into the project's
root directory. When making API changes, make sure to run `npm run start:dev`
to test your changes and generate any new Swagger document bits.

### 📕 Types

We have a couple of scripts to check and adjust missing types.

In order to dry run what types would be added to `package.json`:

```shell
npm run types:auto-check
```

In order to add any missing types to `package.json`:

```shell
npm run types:auto-add
```

### 🚀 Production deployment

A production deployment is a complete build of the project, including the build of the static assets.

```shell
npm run build
```

### 🍕 Pizza service integration

This API integrates with the [pizza service](https://github.com/khulnasoft-opensource/pizza)
to accept requests for ingesting commits to the database.
The environment variables for this are:

```
PIZZA_OVEN_HOST="http://example.com"
PIZZA_OVEN_API=80
```

## 🔑 Database structure

Click the image to see the schema diagram full documentation.

[![er](public/diagrams/schema.svg)](public/diagrams/README.md)

## 🤝 Contributing

We encourage you to contribute to KhulnaSoft OpenSource! Please check out the [Contributing guide](https://docs.khulnasoft.com/contributing/introduction-to-contributing/) for guidelines about how to proceed.

<img align="right" src="https://i.ibb.co/CJfW18H/ship.gif" width="200"/>

## 🍕 Community

Got Questions? Join the conversation in our [Discord](https://discord.gg/U2peSNf23P).  
Find KhulnaSoft OpenSource videos and release overviews on our [YouTube Channel](https://www.youtube.com/channel/UCklWxKrTti61ZCROE1e5-MQ).

## 🎦 Repository Visualization

Below is visual representation of our code repository. It is generated by [Octo Repo Visualizer](https://github.com/githubocto/repo-visualizer).

This visualization is being updated on release to our default branch by our [release workflow](./.github/workflows/release.yml).

[![Visualization of this repository](./public/diagram.svg)
](./src)

## ⚖️ LICENSE

MIT © [KhulnaSoft OpenSource](LICENSE)


CHANGELOG.md
# 📦 khulnasoft-opensource/api changelog

[![conventional commits](https://img.shields.io/badge/conventional%20commits-1.0.0-yellow.svg)](https://conventionalcommits.org)
[![semantic versioning](https://img.shields.io/badge/semantic%20versioning-2.0.0-green.svg)](https://semver.org)

> All notable changes to this project will be documented in this file

## [1.47.0-beta.2](https://github.com/khulnasoft-opensource/api.khulnasoft.com/compare/v1.47.0-beta.1...v1.47.0-beta.2) (2023-08-25)


### 🐛 Bug Fixes

* Highlight users string type in array for swagger ([#280](https://github.com/khulnasoft-opensource/api.khulnasoft.com/issues/280)) ([286e478](https://github.com/khulnasoft-opensource/api.khulnasoft.com/commit/286e47809e36c89148457242a427f4a736f032b2))

## [1.47.0-beta.1](https://github.com/khulnasoft-opensource/api.khulnasoft.com/compare/v1.46.1-beta.1...v1.47.0-beta.1) (2023-08-25)


### 🍕 Features

* add user organizations table migration and entity ([#272](https://github.com/khulnasoft-opensource/api.khulnasoft.com/issues/272)) ([0429526](https://github.com/khulnasoft-opensource/api.khulnasoft.com/commit/042952673b1e932ae8946cc3c915bc61a8ae2326))


### 🐛 Bug Fixes

* update BakeRepoDto.wait to boolean ([#274](https://github.com/khulnasoft-opensource/api.khulnasoft.com/issues/274)) ([45d00e4](https://github.com/khulnasoft-opensource/api.khulnasoft.com/commit/45d00e406196792400ec2d38db8a6923da9a47f5))

### [1.46.1-beta.1](https://github.com/khulnasoft-opensource/api.khulnasoft.com/compare/v1.46.0...v1.46.1-beta.1) (2023-08-25)


### 🐛 Bug Fixes

* Declare "integer" for Swagger types that are not floating points ([#278](https://github.com/khulnasoft-opensource/api.khulnasoft.com/issues/278)) ([9622d88](https://github.com/khulnasoft-opensource/api.khulnasoft.com/commit/9622d882bcf30f1dde78e34a97c305fb2409afc9))

<<<<<<< HEAD
## [1.46.0-beta.10](https://github.com/khulnasoft-opensource/api.khulnasoft.com/compare/v1.46.0-beta.9...v1.46.0-beta.10) (2023-08-24)
=======
## [1.46.0](https://github.com/khulnasoft-opensource/api.khulnasoft.com/compare/v1.45.0...v1.46.0) (2023-08-24)


### 🍕 Features

* Add Swagger YAML ([#266](https://github.com/khulnasoft-opensource/api.khulnasoft.com/issues/266)) ([177882b](https://github.com/khulnasoft-opensource/api.khulnasoft.com/commit/177882b711a4e5b9b231600d786aadc51ba37575))
* add tagged repos to user highlights ([#263](https://github.com/khulnasoft-opensource/api.khulnasoft.com/issues/263)) ([ae5415b](https://github.com/khulnasoft-opensource/api.khulnasoft.com/commit/ae5415b8b9e5b4d2e58f214caf1121746c5507fe))
* remove following users from top users when a user id is provided ([#275](https://github.com/khulnasoft-opensource/api.khulnasoft.com/issues/275)) ([4c7261b](https://github.com/khulnasoft-opensource/api.khulnasoft.com/commit/4c7261b5b8129a223838d3aec7d739355ac08220))
* show reactors who reacted to highlights ([#268](https://github.com/khulnasoft-opensource/api.khulnasoft.com/issues/268)) ([cc2a57c](https://github.com/khulnasoft-opensource/api.khulnasoft.com/commit/cc2a57c6349eb820b20c203a54bdc3faeb495d73))
>>>>>>> main


### 🐛 Bug Fixes

<<<<<<< HEAD
* Declare "integer" for Swagger types that are not floating points ([#278](https://github.com/khulnasoft-opensource/api.khulnasoft.com/issues/278)) ([9622d88](https://github.com/khulnasoft-opensource/api.khulnasoft.com/commit/9622d882bcf30f1dde78e34a97c305fb2409afc9))
=======
* add Swagger API doc fixes for code generation ([#270](https://github.com/khulnasoft-opensource/api.khulnasoft.com/issues/270)) ([1fca4b0](https://github.com/khulnasoft-opensource/api.khulnasoft.com/commit/1fca4b02d1fbcfde38c51e6528a992aa8832b17a))
* add unit test for user-collaboration service and controller ([#267](https://github.com/khulnasoft-opensource/api.khulnasoft.com/issues/267)) ([670e3f9](https://github.com/khulnasoft-opensource/api.khulnasoft.com/commit/670e3f97a0372af6634db252bb6a0624ff949cc1))
* remove extra query parameters from new contributors endpoint ([#265](https://github.com/khulnasoft-opensource/api.khulnasoft.com/issues/265)) ([bffc229](https://github.com/khulnasoft-opensource/api.khulnasoft.com/commit/bffc2294f4bb1d223d0e9f2ffbdf1e828fd450aa))
* update full path to swagger file for writing ([#276](https://github.com/khulnasoft-opensource/api.khulnasoft.com/issues/276)) ([1f3df74](https://github.com/khulnasoft-opensource/api.khulnasoft.com/commit/1f3df7475451904f48f2f3566b4085eb223803bb))
* wrap writing of swagger file ([2b8656d](https://github.com/khulnasoft-opensource/api.khulnasoft.com/commit/2b8656dde65b4c51b7b39a4b79069c15b8ad24be))
>>>>>>> main

## [1.46.0-beta.9](https://github.com/khulnasoft-opensource/api.khulnasoft.com/compare/v1.46.0-beta.8...v1.46.0-beta.9) (2023-08-24)


### 🐛 Bug Fixes

* wrap writing of swagger file ([2b8656d](https://github.com/khulnasoft-opensource/api.khulnasoft.com/commit/2b8656dde65b4c51b7b39a4b79069c15b8ad24be))

## [1.46.0-beta.8](https://github.com/khulnasoft-opensource/api.khulnasoft.com/compare/v1.46.0-beta.7...v1.46.0-beta.8) (2023-08-24)


### 🐛 Bug Fixes

* update full path to swagger file for writing ([#276](https://github.com/khulnasoft-opensource/api.khulnasoft.com/issues/276)) ([1f3df74](https://github.com/khulnasoft-opensource/api.khulnasoft.com/commit/1f3df7475451904f48f2f3566b4085eb223803bb))

## [1.46.0-beta.7](https://github.com/khulnasoft-opensource/api.khulnasoft.com/compare/v1.46.0-beta.6...v1.46.0-beta.7) (2023-08-24)


### 🍕 Features

* remove following users from top users when a user id is provided ([#275](https://github.com/khulnasoft-opensource/api.khulnasoft.com/issues/275)) ([4c7261b](https://github.com/khulnasoft-opensource/api.khulnasoft.com/commit/4c7261b5b8129a223838d3aec7d739355ac08220))

## [1.46.0-beta.6](https://github.com/khulnasoft-opensource/api.khulnasoft.com/compare/v1.46.0-beta.5...v1.46.0-beta.6) (2023-08-23)


### 🍕 Features


src/user/user.entity.ts
import {
  Entity,
  Column,
  BaseEntity,
  PrimaryColumn,
  OneToMany,
  CreateDateColumn,
  DeleteDateColumn,
  UpdateDateColumn,
} from "typeorm";
import { ApiHideProperty } from "@nestjs/swagger";

import {
  ApiModelProperty,
  ApiModelPropertyOptional,
} from "@nestjs/swagger/dist/decorators/api-model-property.decorator";
import { DbRepo } from "../repo/entities/repo.entity";
import { DbRepoToUserVotes } from "../repo/entities/repo.to.user.votes.entity";
import { DbRepoToUserStars } from "../repo/entities/repo.to.user.stars.entity";
import { DbRepoToUserSubmissions } from "../repo/entities/repo.to.user.submissions.entity";
import { DbRepoToUserStargazers } from "../repo/entities/repo.to.user.stargazers.entity";
import { DbInsight } from "../insight/entities/insight.entity";
import { DbUserNotification } from "./entities/user-notification.entity";
import { DbUserHighlight } from "./entities/user-highlight.entity";
import { DbUserHighlightReaction } from "./entities/user-highlight-reaction.entity";
import { DbUserTopRepo } from "./entities/user-top-repo.entity";
import { DbUserCollaboration } from "./entities/user-collaboration.entity";
import { DbUserOrganization } from "./entities/user-organization.entity";

@Entity({ name: "users" })
export class DbUser extends BaseEntity {
  @ApiModelProperty({
    description: "User identifier",
    example: 237133,
    type: "integer",
  })
  @PrimaryColumn("bigint")
  public id!: number;

  @ApiModelProperty({
    description: "Total number of open issues user has across public activity",
    example: 498,
    type: "integer",
  })
  @Column({
    type: "bigint",
    default: 0,
  })
  public open_issues: number;

  @ApiModelPropertyOptional({
    description: "Timestamp representing user creation",
    example: "2016-10-19 13:24:51.000000",
  })
  @CreateDateColumn({
    type: "timestamp without time zone",
    default: () => "now()",
  })
  public created_at?: Date;

  @ApiModelPropertyOptional({
    description: "Timestamp representing user last update",
    example: "2022-08-28 22:04:29.000000",
  })
  @UpdateDateColumn({
    type: "timestamp without time zone",
    default: () => "now()",
  })
  public updated_at?: Date;

  @ApiHideProperty()
  @DeleteDateColumn({
    type: "timestamp without time zone",
    select: false,
  })
  public deleted_at?: Date;

  @ApiHideProperty()
  @Column({
    type: "timestamp without time zone",
    default: () => "to_timestamp(0)",
    select: false,
  })
  public last_fetched_users_at?: Date;

  @ApiModelPropertyOptional({
    description: "Timestamp representing user first open PR",
    example: "2022-08-28 22:04:29.000000",
  })
  @Column({
    type: "timestamp without time zone",
    default: null,
  })
  public first_opened_pr_at?: Date;

  @ApiModelPropertyOptional({
    description: "Timestamp representing user first commit push",
    example: "2022-08-28 22:04:29.000000",
  })
  @Column({
    type: "timestamp without time zone",
    default: null,
  })
  public first_pushed_commit_at?: Date;

  @ApiModelPropertyOptional({
    description: "Timestamp representing user logging in to open khulnasoft for the first time",
    example: "2022-08-28 22:04:29.000000",
  })
  @Column({
    type: "timestamp without time zone",
    default: null,
  })
  public connected_at?: Date;

  @ApiModelProperty({
    description: "User GitHub node id",
    example: "MDQ6VXNlcjIzNzEzMw==",
  })
  @Column({
    type: "character varying",
    length: 20,
  })
  public node_id: string;

  @ApiModelProperty({
    description: "User GitHub avatar URL",
    example: "https://avatars.githubusercontent.com/u/237133?v=4",
  })
  @Column({
    type: "character varying",
    length: 255,
  })
  public avatar_url: string;

  @ApiModelProperty({
    description: "User GitHub gravatar URL",
    example: "",
  })
  @Column({
    type: "character varying",
    length: 255,
  })
  public gravatar_id?: string;

  @ApiModelProperty({
    description: "User GitHub profile URL",
    example: "https://api.github.com/users/0-vortex",
  })
  @Column({
    type: "character varying",
    length: 255,
  })
  public url?: string;

  @ApiModelProperty({
    description: "User unique login name",
    example: "0-vortex",
  })
  @Column({
    type: "character varying",
    length: 255,
  })
  public login: string;

  @ApiModelProperty({
    description: "User email address",
    example: "hello@khulnasoft.com",
  })
  @Column({
    type: "character varying",
    length: 255,
    select: false,
  })
  public email: string;

  @ApiHideProperty()
  @Column({
    default: false,
    select: false,
  })
  public has_stars_data: boolean;

  @ApiModelProperty({
    description: "Flag indicating whether user opted to have a private profile (beta feature",
    example: false,
    default: false,
  })
  @Column({ default: false })
  public is_private: boolean;

  @ApiModelProperty({
    description: "Flag indicating app.khulnasoft user status",
    example: false,
    default: false,
  })
  @Column({ default: false })
  public is_open_khulnasoft_member: boolean;

  @ApiModelProperty({
    description: "Flag indicating user's onboarding status",
    example: false,
    default: false,
  })
  @Column({ default: false })
  public is_onboarded: boolean;

  @ApiModelProperty({
    description: "Flag indicating user's waitlist status",
    example: false,
    default: false,
  })
  @Column({ default: false })
  public is_waitlisted: boolean;

  @ApiModelProperty({
    description: "Insights Role",
    example: 10,
    default: 10,
    type: "integer",
  })
  @Column({ default: 10 })
  public role: number;

  @ApiModelProperty({
    description: "User bio information",
    example: "Khulnasoft User",
  })
  @Column({
    type: "character varying",
    length: 255,
  })
  readonly bio?: string;

  @ApiModelProperty({
    description: "GitHub blog information",
    example: "https://khulnasoft.com/blog",
  })
  @Column({
    type: "character varying",
    length: 255,
  })
  public blog?: string;

  @ApiModelProperty({
    description: "User name information",
    example: "MrPizza",
  })
  @Column({
    type: "character varying",
    length: 255,
  })
  readonly name?: string;

  @ApiModelProperty({
    description: "User Twitter information",
    example: "Khulnasoft",
  })
  @Column({
    type: "character varying",
    length: 15,
  })
  readonly twitter_username?: string;

  @ApiModelProperty({
    description: "LinkedIn URL",
    example: "https://www.linkedin.com/in/brianldouglas",
  })
  @Column({
    type: "character varying",
    length: 255,
  })
  readonly linkedin_url?: string;

  @ApiModelProperty({
    description: "GitHub Sponsors URL",
    example: "https://github.com/sponsors/khulnasoft-opensource",
  })
  @Column({
    type: "character varying",
    length: 255,
  })
  readonly github_sponsors_url?: string;

  @ApiModelProperty({
    description: "Discord URL",
    example: "https://discord.gg/khulnasoft",
  })
  @Column({
    type: "character varying",
    length: 255,
  })
  readonly discord_url?: string;

  @ApiModelProperty({
    description: "User company information",
    example: "Khulnasoft",
  })
  @Column({
    type: "character varying",
    length: 255,
  })
  readonly company?: string;

  @ApiModelProperty({
    description: "User location information",
    example: "San Francisco, CA",
  })
  @Column({
    type: "character varying",
    length: 255,
  })
  readonly location?: string;

  @ApiModelProperty({
    description: "User display local time information",
    example: false,
  })
  @Column({
    type: "boolean",
    default: false,
  })
  readonly display_local_time?: boolean;

  @ApiModelProperty({
    description: "User topic interests",
    example: "javascript",
  })
  @Column({
    type: "character varying",
    length: 200,
  })
  readonly interests?: string;

  @ApiModelProperty({
    description: "GitHub user hireable status",
    example: false,
  })
  @Column({
    type: "boolean",
    default: false,
  })
  public hireable?: boolean;

  @ApiModelProperty({
    description: "GitHub user public repos number",
    example: 0,
    type: "integer",
  })
  @Column({
    type: "bigint",
    default: 0,
  })
  public public_repos: number;

  @ApiModelProperty({
    description: "GitHub user public gists number",
    example: 0,
    type: "integer",
  })
  @Column({
    type: "bigint",
    default: 0,
  })
  public public_gists: number;

  @ApiModelProperty({
    description: "GitHub user public followers number",
    example: 0,
    type: "integer",
  })
  @Column({
    type: "bigint",
    select: false,
    default: 0,
  })
  public followers: number;

  @ApiModelProperty({
    description: "GitHub user public following number",
    example: 0,
    type: "integer",
  })
  @Column({
    type: "bigint",
    select: false,
    default: 0,
  })
  public following: number;

  @ApiModelProperty({
    description: "GitHub user type",
    example: "User",
    default: "User",
  })
  @Column({
    type: "character varying",
    length: 20,
    default: "User",
  })
  public type: string;

  @ApiModelProperty({
    description: "User display public email",
    example: false,
  })
  @Column({
    type: "boolean",
    default: false,
  })
  readonly display_email?: boolean;

  @ApiModelProperty({
    description: "User receives collaboration requests",
    example: false,
  })
  @Column({
    type: "boolean",
    default: false,
  })
  readonly receive_collaboration?: boolean;

  @ApiModelProperty({
    description: "User timezone in UTC",
    example: "UTC-5",
  })
  @Column({
    type: "character varying",
    length: 50,
  })
  readonly timezone?: string;

  @ApiModelProperty({
    description: "GitHub top languages",
    example: "{ TypeScript: 33128, HTML: 453, JavaScript: 90, CSS: 80 }",
    default: {},
  })
  @Column({
    type: "jsonb",
    default: {},
    nullable: false,
  })
  public languages: object;

  @ApiModelProperty({
    description: "User notification count",
    example: 0,
    type: "integer",
  })
  @Column({
    type: "bigint",
    select: false,
    insert: false,
  })
  public notification_count: number;

  @ApiModelProperty({
    description: "User highlights count",
    example: 0,
    type: "integer",
  })
  @Column({
    type: "bigint",
    select: false,
    insert: false,
  })
  public highlights_count: number;

  @ApiModelProperty({
    description: "User following count",
    example: 0,
    type: "integer",
  })
  @Column({
    type: "bigint",
    select: false,
    insert: false,
  })
  public following_count: number;

  @ApiModelProperty({
    description: "User followers count",
    example: 0,
    type: "integer",
  })
  @Column({
    type: "bigint",
    select: false,
    insert: false,
  })
  public followers_count: number;

  @ApiModelProperty({
    description: "Count of user pull requests within the last 30 days",
    example: 0,
    type: "integer",
  })
  @Column({
    type: "bigint",
    select: false,
    insert: false,
  })
  public recent_pull_requests_count: number;

  @ApiModelProperty({
    description: "User average pull request velocity in days over the last 30 days",
    example: 0,
    type: "integer",
  })
  @Column({
    type: "bigint",
    select: false,
    insert: false,
  })
  public recent_pull_request_velocity_count: number;

  @ApiModelProperty({
    description: "Flag to indicate if user is a maintainer of any repos",
    example: true,
  })
  @Column({
    type: "bigint",
    select: false,
    insert: false,
  })
  public is_maintainer: boolean;

  @ApiHideProperty()
  @OneToMany(() => DbInsight, (insight) => insight.user)
  public insights: DbInsight[];

  @ApiHideProperty()
  @OneToMany(() => DbRepo, (repo) => repo.user)
  public repos: DbRepo[];

  @ApiHideProperty()
  @OneToMany(() => DbInsight, (highlights) => highlights.user)
  public highlights: DbUserHighlight[];

  @ApiHideProperty()
  @OneToMany(() => DbUserHighlightReaction, (highlightReactions) => highlightReactions.user)
  public reactions: DbUserHighlightReaction[];

  @ApiHideProperty()
  @OneToMany(() => DbUserCollaboration, (collaboration) => collaboration.user)
  public collaborations: DbUserCollaboration[];

  @ApiHideProperty()
  @OneToMany(() => DbUserCollaboration, (collaboration) => collaboration.request_user)
  public request_collaborations: DbUserCollaboration[];

  @ApiHideProperty()
  @OneToMany(() => DbRepoToUserVotes, (repoToUserVotes) => repoToUserVotes.user)
  public repoToUserVotes: DbRepoToUserVotes[];

  @ApiHideProperty()
  @OneToMany(() => DbRepoToUserStars, (repoToUserStars) => repoToUserStars.user)
  public repoToUserStars: DbRepoToUserStars[];

  @ApiHideProperty()
  @OneToMany(() => DbRepoToUserSubmissions, (repoToUserSubmissions) => repoToUserSubmissions.user)
  public repoToUserSubmissions: DbRepoToUserSubmissions[];

  @ApiHideProperty()
  @OneToMany(() => DbRepoToUserStargazers, (repoToUserStargazers) => repoToUserStargazers.user)
  public repoToUserStargazers: DbRepoToUserStargazers[];

  @ApiHideProperty()
  @OneToMany(() => DbUserTopRepo, (repoToUserTopRepos) => repoToUserTopRepos.user)
  public topRepos: DbUserTopRepo[];

  @ApiHideProperty()
  @OneToMany(() => DbUserNotification, (fromUserNotifications) => fromUserNotifications.from_user)
  public from_user_notifications: DbUserNotification[];

  @ApiHideProperty()
  @OneToMany(() => DbUserOrganization, (userOrgs) => userOrgs.organization_user)
  public organizations: DbUserOrganization[];
}
